File: Minimal/src/Brontide.Minimal.Host/ProviderServingTrustCadence.fs
namespace Brontide.Minimal.Host

open System
open System.Threading
open System.Threading.Tasks

type ProviderServingTrustCadenceSchedule =
    private
    | ProviderServingTrustCadenceSchedule of int * TimeSpan
    member this.MaximumCycles =
        let (ProviderServingTrustCadenceSchedule(value, _)) = this
        value
    member this.Interval =
        let (ProviderServingTrustCadenceSchedule(_, value)) = this
        value

[<RequireQualifiedAccess>]
module ProviderServingTrustCadenceSchedule =
    let create maximumCycles (interval: TimeSpan) =
        if maximumCycles < 1 || maximumCycles > 64 then
            invalidArg (nameof maximumCycles) "A cadence must contain between one and sixty-four cycles."
        if interval <= TimeSpan.Zero || interval > TimeSpan.FromHours 24.0 then
            invalidArg (nameof interval) "A cadence interval must be positive and no greater than twenty-four hours."
        ProviderServingTrustCadenceSchedule(maximumCycles, interval)

type ProviderPublisherTrustPolicyCycle =
    DateTimeOffset -> CancellationToken -> Task<ProviderPublisherTrustPolicyPollResult>

type ProviderServingTrustSweepCycle =
    CancellationToken -> Task<ProviderServingTrustSweepResult option>

[<RequireQualifiedAccess>]
module ProviderServingTrustCycleBinding =
    let policy
        (poller: ProviderPublisherTrustPolicyPoller)
        source
        floorSink
        delay
        : ProviderPublisherTrustPolicyCycle =
        fun now cancellationToken -> poller.PollAsync(source, floorSink, delay, now, cancellationToken)

    let sweep
        registry
        store
        (servingSet: CancellationToken -> Task<ProviderServingActivation list>)
        retirementReason
        : ProviderServingTrustSweepCycle =
        fun cancellationToken -> task {
            let! activations = servingSet cancellationToken
            if List.isEmpty activations then return None
            else
                let! result = ProviderServingTrustSweep.run registry store activations retirementReason
                return Some result
        }

/// Every code a trust cycle can return, with whether the cadence may continue after it. Producers use
/// these values rather than literals and CBI48's journal validates against this set, so a new code
/// cannot be produced without the journal knowing it — which is what CBI61's two additions did before
/// this was one vocabulary.
///
/// `establishes` carries the second question a consumer asks of a committed observation: did the cycle
/// reporting this code establish current policy? It is stated here rather than in the consumer for the
/// same reason `continues` is, one consumer over — a code a cycle can produce and a consumer cannot
/// classify is exactly the seam defect CBI62 found. It is `None` for a non-continuing code, because
/// such a code makes the run terminal in the write that commits it and so never precedes another
/// observation; `Stopped` is the one that could not be answered anyway, since a cycle produces it both
/// for a poll that was not current and for a current poll whose sweep failed.
[<RequireQualifiedAccess>]
module ProviderServingTrustCycleCodes =
    [<Literal>]
    let Current = "provider-trust-cycle-current"
    [<Literal>]
    let Withdrawn = "provider-trust-cycle-withdrawn"
    [<Literal>]
    let Stopped = "provider-trust-cycle-stopped"
    [<Literal>]
    let Canceled = "provider-trust-cycle-canceled"
    [<Literal>]
    let AuthorityBehind = "provider-trust-cycle-authority-behind"
    [<Literal>]
    let AuthorityUnretained = "provider-trust-cycle-authority-unretained"
    [<Literal>]
    let Offline = "provider-trust-cycle-offline"

    let private codes =
        Map.ofList
            [ Current, (true, Some true)
              Withdrawn, (true, Some true)
              Stopped, (false, None)
              Canceled, (false, None)
              AuthorityBehind, (false, None)
              AuthorityUnretained, (false, None)
              Offline, (true, Some false) ]

    let all = codes |> Map.toList |> List.map fst

    let isKnown code = codes |> Map.containsKey code

    let continues code = codes |> Map.tryFind code |> Option.map fst |> Option.defaultValue false

    /// Whether a cycle reporting this code established current policy, or `None` when the vocabulary
    /// does not answer it. A consumer that meets a `None` must refuse rather than choose a side.
    let establishes code = codes |> Map.tryFind code |> Option.bind snd

/// What one cycle's availability evaluation decided and what enforcing it did. The cycle reports the
/// CBI49 decision and CBI50's counts rather than CBI50's own result, which keeps this result free of
/// the enforcement component's types; per-member evidence stays where its ordering and
/// failure-isolation properties are pinned. `DecisionCode` is `None` when CBI50 refused the serving
/// snapshot before any policy was evaluated.
type ProviderTrustCycleAvailability =
    { EnforcementCode: string
      DecisionCode: string option
      Deadline: DateTimeOffset option
      RetryAt: DateTimeOffset option
      AdmittedCount: int
      StoppedCount: int }
    member this.PermitsContinuation =
        this.DecisionCode = Some "offline-existing-service"
        || this.DecisionCode = Some "offline-idle"

/// One cycle's observations. `Rotation` is `None` for a cadence that does not govern authority
/// rotation, which is every cadence composed before CBI61; a governed cycle carries the CBI60 result
/// it ran before the poll, and `Poll` is `None` when that rotation stopped the cycle before the
/// policy endpoint was contacted. `Availability` is `None` for a cadence that governs no availability
/// policy, which is every cadence composed before CBI64, and for a cycle that established current
/// policy or never reached the endpoint.
type ProviderServingTrustCycleResult =
    { Code: string
      Poll: ProviderPublisherTrustPolicyPollResult option
      Sweep: ProviderServingTrustSweepResult option
      ServingCount: int
      Rotation: ProviderPolicyAuthorityCycleResult option
      Availability: ProviderTrustCycleAvailability option }
    member this.CanContinue = ProviderServingTrustCycleCodes.continues this.Code
    member this.IsCanceled = this.Code = ProviderServingTrustCycleCodes.Canceled

type ProviderServingTrustCycle =
    DateTimeOffset -> CancellationToken -> Task<ProviderServingTrustCycleResult>

[<RequireQualifiedAccess>]
module ProviderServingTrustCycle =
    let create
        (policy: ProviderPublisherTrustPolicyCycle)
        (serving: ProviderServingTrustSweepCycle)
        : ProviderServingTrustCycle =
        fun now cancellationToken -> task {
            let! poll = policy now cancellationToken
            if poll.Code = "policy-poll-canceled" then
                return
                    { Code = ProviderServingTrustCycleCodes.Canceled; Poll = Some poll
                      Sweep = None; ServingCount = 0; Rotation = None; Availability = None }
            elif not poll.IsCurrent then
                return
                    { Code = ProviderServingTrustCycleCodes.Stopped; Poll = Some poll
                      Sweep = None; ServingCount = 0; Rotation = None; Availability = None }
            else
                let! sweep = serving cancellationToken
                match sweep with
                | None ->
                    return
                        { Code = ProviderServingTrustCycleCodes.Current; Poll = Some poll
                          Sweep = None; ServingCount = 0; Rotation = None; Availability = None }
                | Some result ->
                    let code =
                        match result.Code with
                        | "serving-trust-sweep-current" -> ProviderServingTrustCycleCodes.Current
                        | "serving-trust-sweep-withdrawn" -> ProviderServingTrustCycleCodes.Withdrawn
                        | _ -> ProviderServingTrustCycleCodes.Stopped
                    return
                        { Code = code; Poll = Some poll; Sweep = Some result
                          ServingCount = result.Members.Length; Rotation = None; Availability = None }
        }

type ProviderServingTrustCadenceDelay =
    DateTimeOffset -> TimeSpan -> CancellationToken -> Task<DateTimeOffset>

type ProviderServingTrustCadenceCycle =
    { Instant: DateTimeOffset
      Result: ProviderServingTrustCycleResult }

type ProviderServingTrustCadenceResult =
    { Code: string
      Cycles: ProviderServingTrustCadenceCycle list
      Gaps: TimeSpan list }

[<RequireQualifiedAccess>]
module ProviderServingTrustCadence =
    let run
        (schedule: ProviderServingTrustCadenceSchedule)
        (cycle: ProviderServingTrustCycle)
        (delay: ProviderServingTrustCadenceDelay)
        start
        (cancellationToken: CancellationToken)
        = task {
            if isNull (box cycle) then nullArg (nameof cycle)
            if isNull (box delay) then nullArg (nameof delay)

            let cycles = ResizeArray<ProviderServingTrustCadenceCycle>()
            let gaps = ResizeArray<TimeSpan>()
            let mutable instant = start
            let mutable code: string option = None

            while cycles.Count < schedule.MaximumCycles && Option.isNone code do
                if cancellationToken.IsCancellationRequested then
                    code <- Some "provider-trust-cadence-canceled"
                else
                    let! canRun = task {
                        if cycles.Count = 0 then return true
                        else
                            try
                                let! next = delay instant schedule.Interval cancellationToken
                                if cancellationToken.IsCancellationRequested then return false
                                else
                                    if next <= instant then
                                        invalidOp "A cadence delay must advance the cycle instant."
                                    instant <- next
                                    gaps.Add schedule.Interval
                                    return true
                            with :? OperationCanceledException -> return false
                    }
                    if not canRun then code <- Some "provider-trust-cadence-canceled"
                    else
                        let! result = cycle instant cancellationToken
                        cycles.Add { Instant = instant; Result = result }
                        if result.IsCanceled then code <- Some "provider-trust-cadence-canceled"
                        elif not result.CanContinue then code <- Some "provider-trust-cadence-stopped"

            return
                { Code = code |> Option.defaultValue "provider-trust-cadence-complete"
                  Cycles = List.ofSeq cycles; Gaps = List.ofSeq gaps }
        }

File: Reference/src/Brontide.Reference.Studio/ProviderServingTrustCadence.cs
namespace Brontide.Reference.Studio;

public sealed record ProviderServingTrustCadenceSchedule
{
    private ProviderServingTrustCadenceSchedule(int maximumCycles, TimeSpan interval)
    {
        MaximumCycles = maximumCycles;
        Interval = interval;
    }

    public int MaximumCycles { get; }
    public TimeSpan Interval { get; }

    public static ProviderServingTrustCadenceSchedule Create(int maximumCycles, TimeSpan interval)
    {
        if (maximumCycles is < 1 or > 64) throw new ArgumentOutOfRangeException(nameof(maximumCycles));
        if (interval <= TimeSpan.Zero || interval > TimeSpan.FromHours(24))
            throw new ArgumentOutOfRangeException(nameof(interval));
        return new(maximumCycles, interval);
    }
}

public interface IProviderPublisherTrustPolicyCycle
{
    Task<ProviderPublisherTrustPolicyPollResult> PollAsync(
        DateTimeOffset now,
        CancellationToken cancellationToken);
}

/// <summary>Binds one CBI41 poller to the endpoint, floor, and delay owned by its host.</summary>
public sealed class ProviderPublisherTrustPolicyCycle(
    ProviderPublisherTrustPolicyPoller poller,
    IProviderPublisherTrustPolicyDistributionSource source,
    IProviderPublisherTrustPolicyFloorSink floorSink,
    IProviderPublisherTrustPolicyPollDelay delay) : IProviderPublisherTrustPolicyCycle
{
    public Task<ProviderPublisherTrustPolicyPollResult> PollAsync(
        DateTimeOffset now,
        CancellationToken cancellationToken) =>
        poller.PollAsync(source, floorSink, delay, now, cancellationToken);
}

public interface IProviderServingTrustSweepCycle
{
    /// <summary>Returns null only when the current serving-set snapshot is empty.</summary>
    ValueTask<ProviderServingTrustSweepResult?> SweepAsync(CancellationToken cancellationToken);
}

/// <summary>Snapshots the serving set after policy polling and invokes CBI46 when it is non-empty.</summary>
public sealed class ProviderServingTrustSweepCycle(
    DurableProviderPublisherTrustPolicyRegistry registry,
    ContentAddressedProviderStore store,
    Func<CancellationToken, ValueTask<IReadOnlyList<ProviderServingActivation>>> servingSet,
    string retirementReason) : IProviderServingTrustSweepCycle
{
    public async ValueTask<ProviderServingTrustSweepResult?> SweepAsync(CancellationToken cancellationToken)
    {
        var activations = await servingSet(cancellationToken).ConfigureAwait(false);
        ArgumentNullException.ThrowIfNull(activations);
        if (activations.Count == 0) return null;
        return await ProviderServingTrustSweep.RunAsync(
            registry, store, activations, retirementReason, cancellationToken).ConfigureAwait(false);
    }
}

/// <summary>
/// Every code a trust cycle can return, with whether the cadence may continue after it. Producers
/// use these constants rather than literals and CBI48's journal validates against this set, so a new
/// code cannot be produced without the journal knowing it — which is what CBI61's two additions did
/// before this was one vocabulary.
/// </summary>
public static class ProviderServingTrustCycleCodes
{
    public const string Current = "provider-trust-cycle-current";
    public const string Withdrawn = "provider-trust-cycle-withdrawn";
    public const string Stopped = "provider-trust-cycle-stopped";
    public const string Canceled = "provider-trust-cycle-canceled";
    public const string AuthorityBehind = "provider-trust-cycle-authority-behind";
    public const string AuthorityUnretained = "provider-trust-cycle-authority-unretained";

    private static readonly Dictionary<string, bool> Continuing = new(StringComparer.Ordinal)
    {
        [Current] = true,
        [Withdrawn] = true,
        [Stopped] = false,
        [Canceled] = false,
        [AuthorityBehind] = false,
        [AuthorityUnretained] = false,
    };

    public static IReadOnlyCollection<string> All => Continuing.Keys;

    public static bool IsKnown(string code) => code is not null && Continuing.ContainsKey(code);

    public static bool Continues(string code) =>
        code is not null && Continuing.TryGetValue(code, out var value) && value;
}

/// <summary>
/// One cycle's observations. <paramref name="Rotation"/> is absent for a cadence that does not
/// govern authority rotation, which is every cadence composed before CBI61; a governed cycle carries
/// the CBI60 result it ran before the poll, and <paramref name="Poll"/> is absent when that rotation
/// stopped the cycle before the policy endpoint was contacted.
/// </summary>
public sealed record ProviderServingTrustCycleResult(
    string Code,
    ProviderPublisherTrustPolicyPollResult? Poll,
    ProviderServingTrustSweepResult? Sweep,
    int ServingCount,
    ProviderPolicyAuthorityCycleResult? Rotation = null)
{
    public bool CanContinue => ProviderServingTrustCycleCodes.Continues(Code);
    public bool IsCanceled => Code == ProviderServingTrustCycleCodes.Canceled;
}

public interface IProviderServingTrustCycle
{
    Task<ProviderServingTrustCycleResult> RunAsync(
        DateTimeOffset now,
        CancellationToken cancellationToken);
}

/// <summary>Composes one current-policy poll with at most one current-serving-set sweep.</summary>
public sealed class ProviderServingTrustCycle(
    IProviderPublisherTrustPolicyCycle policy,
    IProviderServingTrustSweepCycle serving) : IProviderServingTrustCycle
{
    public async Task<ProviderServingTrustCycleResult> RunAsync(
        DateTimeOffset now,
        CancellationToken cancellationToken)
    {
        var poll = await policy.PollAsync(now, cancellationToken).ConfigureAwait(false);
        if (poll.Code == "policy-poll-canceled")
            return new(ProviderServingTrustCycleCodes.Canceled, poll, null, 0);
        if (!poll.IsCurrent)
            return new(ProviderServingTrustCycleCodes.Stopped, poll, null, 0);

        var sweep = await serving.SweepAsync(cancellationToken).ConfigureAwait(false);
        if (sweep is null)
            return new(ProviderServingTrustCycleCodes.Current, poll, null, 0);
        return sweep.Code switch
        {
            "serving-trust-sweep-current" =>
                new(ProviderServingTrustCycleCodes.Current, poll, sweep, sweep.Members.Count),
            "serving-trust-sweep-withdrawn" =>
                new(ProviderServingTrustCycleCodes.Withdrawn, poll, sweep, sweep.Members.Count),
            _ => new(ProviderServingTrustCycleCodes.Stopped, poll, sweep, sweep.Members.Count),
        };
    }
}

public interface IProviderServingTrustCadenceDelay
{
    Task<DateTimeOffset> DelayAsync(
        DateTimeOffset now,
        TimeSpan duration,
        CancellationToken cancellationToken);
}

public sealed record ProviderServingTrustCadenceCycle(
    DateTimeOffset Instant,
    ProviderServingTrustCycleResult Result);

public sealed record ProviderServingTrustCadenceResult(
    string Code,
    IReadOnlyList<ProviderServingTrustCadenceCycle> Cycles,
    IReadOnlyList<TimeSpan> Gaps);

/// <summary>Runs a bounded host cadence without an ambient timer or hidden continuation.</summary>
public sealed class ProviderServingTrustCadence(ProviderServingTrustCadenceSchedule schedule)
{
    public async Task<ProviderServingTrustCadenceResult> RunAsync(
        IProviderServingTrustCycle cycle,
        IProviderServingTrustCadenceDelay delay,
        DateTimeOffset start,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(cycle);
        ArgumentNullException.ThrowIfNull(delay);

        var cycles = new List<ProviderServingTrustCadenceCycle>();
        var gaps = new List<TimeSpan>();
        var instant = start;

        while (cycles.Count < schedule.MaximumCycles)
        {
            if (cancellationToken.IsCancellationRequested)
                return new("provider-trust-cadence-canceled", cycles, gaps);

            if (cycles.Count > 0)
            {
                DateTimeOffset next;
                try
                {
                    next = await delay.DelayAsync(instant, schedule.Interval, cancellationToken)
                        .ConfigureAwait(false);
                }
                catch (OperationCanceledException)
                {
                    return new("provider-trust-cadence-canceled", cycles, gaps);
                }
                if (cancellationToken.IsCancellationRequested)
                    return new("provider-trust-cadence-canceled", cycles, gaps);
                if (next <= instant)
                    throw new InvalidOperationException("A cadence delay must advance the cycle instant.");
                instant = next;
                gaps.Add(schedule.Interval);
            }

            var result = await cycle.RunAsync(instant, cancellationToken).ConfigureAwait(false);
            cycles.Add(new(instant, result));
            if (result.IsCanceled)
                return new("provider-trust-cadence-canceled", cycles, gaps);
            if (!result.CanContinue)
                return new("provider-trust-cadence-stopped", cycles, gaps);
        }

        return new("provider-trust-cadence-complete", cycles, gaps);
    }
}
